TheWidlarzGroup/react-native-video
# Video caching on Android

## Description

Video caching might be great for users who play the same video multiple times or would like to play that same video offline. Currently `react-native-video` doesn't provide a way to cache videos on Android. There has been a few changes to ExoPlayer which makes video caching simpler than before. This PR intends to add caching in a simple and effective way. 

These changes wrap the default ExoPlayer `DataSource` in to a `CacheDataSource`. When there's no cache configured, the default data source is used and not wrapped. If cache is enabled, the video will be retrieved from the default DataSoure on a cache MISS. On a cache HIT, the video is retrieved from the file system.

## PR todos

- [x] Provide an example of how to test the change
- [x] Focus the PR on only one area
- [x] Describe the changes

## How to test

1. Deploy debug app to real device or emulator
2. Connect to instance and open shell by running command `adb shell`
3. Open app environment by running command in opened shell `run-as <package-name>`
4. Check cache size `du -h cache/videos` and/or see that's empty
5. Initialize video player with cache configuration
```
<Video
    {...}
    cache={{
        dir: "videos",
        maxSizeBytes: 100 * 1024 * 1024,
    }}
/>
```
6. Play a video with that same player
7. Check again cache size `du -h cache/videos` and validate that it contains more data
8. Play more videos until 100 MB data is cached
9. Check again cache size `du -h cache/videos` and validate that it does not contain more than 100MB data
